### PR TITLE
Adds linked editing for JSX tags

### LINKED ISSUE #51832: Mirror Cursor for JSX
microsoft/vscode#47069 implements this for HTML.

We should look for a way to support this for JSX/TSX.

### PR DESCRIPTION
**Pull request by iisaduan**
Adds linked editing of JSX tags (mirror cursor) 

fixes #51832

### CODE REVIEW

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -89,6 +89,7 @@ declare namespace ts {
         namespace protocol {
             enum CommandTypes {
                 JsxClosingTag = "jsxClosingTag",
+                JsxMirrorCursor = "jsxMirrorCursor",
```
Comment (mjbvz) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:92]: If we want to align this with the LSP, we may instead want to call this `jsxLinkedEditingRange`

**Review by mjbvz — CHANGES_REQUESTED** [review]

File: tests/cases/fourslash/jsxTagLinkedEdit7.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+////);
+////const jsx3 = (
+////    </*6*/>
+////        <div>
+////    </>
+////);
+
+const linkedCursors7 = {ranges: [{start: 14, end: 14}, 
+                            {start: 43, end: 43}],
+                        wordPattern : undefined};
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/jsxTagLinkedEdit7.ts:21]: The way testing works right now is not ideal because we can't tell where positions 14 and 43 are. Instead, consider just handing over an array of marker names.

File: tests/cases/fourslash/jsxTagLinkedEdit7.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+////    <>/*5*/
+////    </>
+////);
+////const jsx3 = (
+////    </*6*/>
+////        <div>
+////    </>
+////);
+
+const linkedCursors7 = {ranges: [{start: 14, end: 14},
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/jsxTagLinkedEdit7.ts:19]: The data here is currently incorrect because of the `pos + 1` bug I mentioned earlier. You need to make sure you get the start, not-including trivia (i.e. whitespace + comments).

File: src/services/services.ts [hunk 4]
```diff
@@ -2478,6 +2486,58 @@ export function createLanguageService(
...
+                        if ((isJSXTagName(n) && !isJsxSelfClosingElement(n.parent))
+                            || (n.kind===SyntaxKind.LessThanToken && isJSXTagName(findNextToken(n,n.parent,sourceFile) ?? n))
+                            || n.kind===SyntaxKind.SlashToken) {
+                            return true;
+                        }
+                        return "quit";
+                    }
+                    return false;
+                }
+                )?.parent as JsxOpeningElement | JsxClosingElement | undefined;
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2525]: This is super hard to read - can you move the result of `findAncestor` to its own variable?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Looks like this is all coming together pretty nicely. I left a bit of feedback about

- why things weren't working correctly with fragments,
- some ideas for how fourslash testing can be made easier
- some thoughts on the API (`TextSpan` vs. `TextRange` - maybe we revisit that?)
- some style feedback

With regards to style, at some point we should consider getting an automatic formatter - but in the meantime, try to stay consistent with the rest of the codebase, and use the editor's built-in formatter as well.

File: tests/cases/fourslash/jsxTagLinkedEdit1.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+////   </di/*A*/v>
+////);
+
+// @Filename: /whitespaceClosing.tsx
+////const jsx = (
+////   <di/*B*/v>
+////   <//*15*/ /*16*/div/*17*/ /*18*/> /*19*/
+////);
+
+const linkedCursors1 = {ranges: [{start: 19, length: 3},
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/jsxTagLinkedEdit1.ts:33]: The goal of markers/ranges in fourslash is to avoid hardcoded file positions like this in testcases, since they're very tedious to maintain.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by iisaduan** [comment]
I've renamed the feature to `linkedEditing...`, to match the LSP and allow for potential non-JSX tag usage in the future.

File: src/services/services.ts [hunk 6]
```diff
@@ -2478,6 +2482,51 @@ export function createLanguageService(
...
+                        }
+                        return "quit";
+                    }
+                    return false;
+                });
+            if (!tag) return undefined;
+
+            const element = tag.parent as JsxOpeningElement | JsxClosingElement;
+            const openTag = { start: element.parent.openingElement.tagName.getStart(sourceFile), end: element.parent.openingElement.tagName.end };
+            const endTag = { start: element.parent.closingElement.tagName.getStart(sourceFile), end: element.parent.closingElement.tagName.end };
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2518]: `openTag` and `endTag` are never used except just to "namespace" some variables. I would just create `openTagStart`, `openTagEnd`, `closingTagStart`, and `closingTagEnd`.

File: src/services/services.ts [hunk 7]
```diff
@@ -2478,6 +2482,51 @@ export function createLanguageService(
...
+
+            // only return a mirror if the cursor is within a tag name
+            if (!(openTag.start <= position && position <= openTag.end || endTag.start <= position && position <= endTag.end)) return undefined;
+            if (element.parent.openingElement.tagName.getText(sourceFile) !== element.parent.closingElement.tagName.getText(sourceFile)) return undefined;
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2522]: leave a quick comment for this line on what you're testing here (as you did above)

File: src/services/services.ts [hunk 8]
```diff
@@ -2478,6 +2482,51 @@ export function createLanguageService(
...
+        if (isJsxFragment(token.parent.parent)) {
+            const openPos = token.parent.parent.openingFragment.getStart(sourceFile) + 1; // "<".length
+            const closePos = token.parent.parent.closingFragment.getStart(sourceFile) + 2; // "</".length
+
+            // only allows mirroring right after opening bracket: <| ></| >
+            if ((position !== openPos) && (position !== closePos)) return undefined;
+
+            const ranges = [{ start: openPos, length: 0 }, { start: closePos, length: 0 }];
+            const wordPattern = undefined;
+            return { ranges, wordPattern };
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2499]: `wordPattern` is optional, so you can omit it.

File: src/services/services.ts [hunk 9]
```diff
@@ -2478,6 +2482,51 @@ export function createLanguageService(
...
+            const element = tag.parent as JsxOpeningElement | JsxClosingElement;
+            const openTag = { start: element.parent.openingElement.tagName.getStart(sourceFile), end: element.parent.openingElement.tagName.end };
+            const endTag = { start: element.parent.closingElement.tagName.getStart(sourceFile), end: element.parent.closingElement.tagName.end };
+
+            // only return a mirror if the cursor is within a tag name
+            if (!(openTag.start <= position && position <= openTag.end || endTag.start <= position && position <= endTag.end)) return undefined;
+            if (element.parent.openingElement.tagName.getText(sourceFile) !== element.parent.closingElement.tagName.getText(sourceFile)) return undefined;
+
+            const ranges = [{ start: openTag.start, length: openTag.end - openTag.start }, { start: endTag.start, length: endTag.end - endTag.start }];
+            const wordPattern = element.tagName.getText(sourceFile);
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2525]: Maybe capture the opening tag text and reuse it here so you don't slice twice.

File: src/services/services.ts [hunk 10]
```diff
@@ -2478,6 +2482,51 @@ export function createLanguageService(
...
+            // only allows mirroring right after opening bracket: <| ></| >
+            if ((position !== openPos) && (position !== closePos)) return undefined;
+
+            const ranges = [{ start: openPos, length: 0 }, { start: closePos, length: 0 }];
+            const wordPattern = undefined;
+            return { ranges, wordPattern };
+        }
+        else {
+            // looks for mirroring in element tag names
+            const tag = findAncestor(token,
```
Comment (DanielRosenwasser) [inline src/services/services.ts:2503]: If you feed in `token.parent` instead of `token` itself, the entire body can get rid of all the `parent.parent` references and talk only about `n`. Then you don't have to create a separate `element` variable below.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
I think you could use a few more edge-case tests. Tests where you're at the beginning of the file, end of the file, and where the code is not syntactically valid (e.g. a closing tag missing its tag name, or a closing fragment tag that "accidentally" has a tag name.

File: tests/cases/fourslash/linkedEditingJsxTag6.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+/// <reference path='fourslash.ts' />
+
+// @Filename: /namespace.tsx
+////const jsx = (
+////    </*start*/someNamespa/*3*/ce./*2*/Thing>
+////    <//*end*/someNamespace/*1*/.Thing>
```
[the remark this answers is not in the document]
Comment (iisaduan) [reply tests/cases/fourslash/linkedEditingJsxTag6.ts:6]: Interesting! @mjbvz, should we add linked editing in these cases?
I could see there being two ranges in each of these three cases: a cursor in the `foo` of the opening tag linking to the `foo` in the closing tag, and the same with `bar`, for example: 
```
const jsx1 = <foo     .   |bar> </foo.|bar>
```
but not 
```
const jsx1 = <foo     .|   bar> </foo.bar>
```

I tested the current behavior of all of the above cases, and currently, the code does not return a linked cursor when there's trivia in the middle of an access expression tag name. We do, however, have linking in the following: 
```
const jsx = </* comment */ |div /* comment */> <//* comment */ |div /* comment */>
```

File: src/server/session.ts [hunk 12]
```diff
@@ -3644,6 +3657,20 @@ function positionToLineOffset(info: ScriptInfoOrConfig, position: number): proto
     return isConfigFile(info) ? locationFromLineAndCharacter(info.getLineAndCharacterOfPosition(position)) : info.positionToLineOffset(position);
 }
 
+function convertLinkedEditInfoToRanges(linkedEdit: LinkedEditingInfo, scriptInfo: ScriptInfo): protocol.LinkedEditingRanges {
+    return {
+        ranges: linkedEdit.ranges.map(
+            s => {
+                return {
+                start: scriptInfo.positionToLineOffset(s.start),
+                end: scriptInfo.positionToLineOffset(s.start + s.length)
```
Comment (DanielRosenwasser) [inline src/server/session.ts:3666]:
```suggestion
                    start: scriptInfo.positionToLineOffset(s.start),
                    end: scriptInfo.positionToLineOffset(s.start + s.length),
```

File: tests/cases/fourslash/linkedEditingJsxTag6.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply tests/cases/fourslash/linkedEditingJsxTag6.ts:6]: That would be kind of cool- but I think for a minimum viable product, it's okay if you just say the tag names have to be the same verbatim.

That still has some weird behavior, like linked cursors being available throughout the following example:

```ts
let jsx4 =
    <foo   .// hi
    bar>
   </foo   .// hi
    bar>
```

But it's probably fine.

File: src/services/services.ts [hunk 13]
```diff
@@ -2478,6 +2482,56 @@ export function createLanguageService(
...
+            // only return linked cursors if the cursor is within a tag name
+            if (!(openTagStart <= position && position <= openTagEnd || closeTagStart <= position && position <= closeTagEnd)) return undefined;
+
+            // only return linked cursors if text in both tags is identical
+            const openingTagText = element.parent.openingElement.tagName.getText(sourceFile);
+            if (openingTagText !== element.parent.closingElement.tagName.getText(sourceFile)) return undefined;
+
+            return {
+                ranges: [{ start: openTagStart, length: openTagEnd - openTagStart }, { start: closeTagStart, length: closeTagEnd - closeTagStart }],
+                wordPattern: openingTagText
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/services/services.ts:2530]: What happens if the returned ranges don't satisfy the regex? No linked cursors I assume?

I think the problem with that regex is that it expects at least two characters. Maybe you wanted:

```
[a-z\._$][a-z0-9\._$]*
```

?

[the remark this answers is not in the document]
Comment (mjbvz) [reply src/services/services.ts:2530]: Yes, but also try omitting it and seeing if VS Code does the right thing already. You should only need to return `wordRange` if the default behavior isn't correct

[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/services/services.ts:2530]: Yeah, in that case let's leave off `wordPattern` for now.

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 14]
```diff
@@ -877,6 +878,16 @@ declare namespace ts {
             interface JsxClosingTagResponse extends Response {
                 readonly body: TextInsertion;
             }
+            interface LinkedEditingRequest extends FileLocationRequest {
+                readonly command: CommandTypes.LinkedEditing;
+            }
+            interface LinkedEditingRanges {
```
[the remark this answers is not in the document]
Comment (iisaduan) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:884]: I had written the `LinkedEditingRanges` and `LinkedEditingRangeResponse` this way, as this is how they were set up in the stubs and LSP

File: tests/cases/fourslash/fourslash.ts [hunk 15]
```diff
@@ -759,6 +760,11 @@ declare namespace FourSlashInterface {
         generateReturnInDocTemplate?: boolean;
     }
 
+    type LinkedEditingInfo = {
+        readonly ranges : { start: number, length: number }[];
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/fourslash.ts:764]:
```suggestion
        readonly ranges: { start: number, length: number }[];

```

File: tests/cases/fourslash/fourslash.ts [hunk 16]
```diff
@@ -759,6 +760,11 @@ declare namespace FourSlashInterface {
         generateReturnInDocTemplate?: boolean;
     }
 
+    type LinkedEditingInfo = {
+        readonly ranges : { start: number, length: number }[];
+        wordPattern? : string;
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/fourslash.ts:765]:
```suggestion
        wordPattern?: string;

```

File: src/services/services.ts [hunk 17]
```diff
@@ -2478,6 +2481,57 @@ export function createLanguageService(
...
+                    if (!n.parent) return "quit";
+                    else if (isJsxElement(n.parent)) {
+                        if (isJsxOpeningElement(n) || isJsxClosingElement(n)) {
+                            return true;
+                        }
+                        return "quit";
+                    }
+                    return false;
+                });
+            if (!element || !(isJsxOpeningElement(element) || isJsxClosingElement(element))) return undefined;
```
Comment (RyanCavanaugh) [inline src/services/services.ts:2515]: Is `(isJsxOpeningElement(element) || isJsxClosingElement(element))` possibly `false` given what the `findAncestor` call does? Fine to leave in just in case, but I'm curious if this/how this happens

**Review by RyanCavanaugh — APPROVED** [review]

File: tests/baselines/reference/api/tsserverlibrary.d.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (jakebailey) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:884]: I would also prefer if it matched the style we had in tsserver and was named `Body`; LSP may define it one way but I don't think we should be inconsistent to match them.

File: tests/cases/fourslash/linkedEditingJsxTag4.ts [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+const startPos = test.markerByName("0").position;
+const endPos =  test.markerByName("6").position;
+const linkedCursors = {
+    ranges: [{ start: startPos, length: 3 }, { start: endPos, length: 3 }],
+};
+
+verify.linkedEditing( {
+    "0": linkedCursors,
+    "1": linkedCursors,
+    "2": undefined,
```
Comment (jakebailey) [inline tests/cases/fourslash/linkedEditingJsxTag4.ts:37]: Maybe this is a lot of work, but maybe it's better to have a baseline for this instead, rather than all of these `undefined`s? I know that a lot of our current tests have this, but, I think we're trying to make them friendlier, e.g. @sandersn's new completion baselines.

File: src/services/services.ts [hunk 19]
```diff
@@ -2478,6 +2481,57 @@ export function createLanguageService(
...
+            // only allows linked editing right after opening bracket: <| ></| >
+            if ((position !== openPos) && (position !== closePos)) return undefined;
+
+            return { ranges: [{ start: openPos, length: 0 }, { start: closePos, length: 0 }] };
+        }
+        else {
+            // determines if the cursor is in an element tag
+            const element = findAncestor(token.parent,
+                n => {
+                    if (!n.parent) return "quit";
```
Comment (jakebailey) [inline src/services/services.ts:2506]: Is this required? I think the only case where you won't have a parent is on `SourceFile`, in which case iteration is going to stop.

File: src/services/services.ts [hunk 20]
```diff
@@ -2478,6 +2481,57 @@ export function createLanguageService(
...
+                        if (isJsxOpeningElement(n) || isJsxClosingElement(n)) {
+                            return true;
+                        }
+                        return "quit";
+                    }
+                    return false;
+                });
+            if (!element || !(isJsxOpeningElement(element) || isJsxClosingElement(element))) return undefined;
+
+            const openTagStart = element.parent.openingElement.tagName.getStart(sourceFile);
```
Comment (jakebailey) [inline src/services/services.ts:2517]: `element.parent.openingElement` and `element.parent.closingElement` seem to be getting used a lot; maybe pull them out into a constant to so we aren't doing the full chain over and over?

**Review by jakebailey — COMMENTED** [review]
Things seem good, but, I think I would like to see the protocol naming be consistent, as once we add it, we're not going to change it.

File: src/services/services.ts (hunk 17 above)
Comment (iisaduan) [reply src/services/services.ts:2515]: Seems to now be mostly for the typeguard feature, as it seems other code I've added since covers these cases. Should I change it to an `as JsxOpeningElement | JsxClosingElement | undefined` assertion to the result of `findAncestor`?

**Review by jakebailey — APPROVED** [review]

File: src/services/services.ts (hunk 19 above)
Comment (iisaduan) [reply src/services/services.ts]: It caused an issue when you checked the kind of `.parent`, and `.parent` is `undefined`, but turns out, I was able to rewrite `findAncestor` so I didn't need to check `.parent` anymore
